### PR TITLE
PSR-43 created test cases for views and forms

### PR DESCRIPTION
**Pull request by narendrapaduchuri**

### CODE REVIEW

File: app/forms/DatePageFormProvider.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+package forms
+
+import forms.mappings.Mappings
+import play.api.data.Form
+
+import java.time.LocalDate
+import javax.inject.Inject
+
+class DatePageFormProvider @Inject() extends Mappings {
```
Comment (kurtlogan) [inline app/forms/DatePageFormProvider.scala:25]: I was thinking that maybe the provider could return a case class that wraps the form and the view can accept that as a parameter. This would mean that a dev couldn't create a form with the wrong fields but still satisfys the type signature. This will need some thought as form as methods that are needed and would break it were to be a parameter on another case class. Perhaps we can go through this

File: app/viewmodels/models/DatePageViewModel.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package viewmodels
+
+import play.api.i18n.Messages
+import play.api.mvc.Call
+
+case class DisplayMessage(key: String, args: List[Any]) {
```
Comment (kurtlogan) [inline app/viewmodels/models/DatePageViewModel.scala:22]: Lets pull display message into it's own class

File: test/controllers/SampleTestSpec.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class SampleTestSpec extends AnyWordSpec with Matchers{
```
Comment (kurtlogan) [inline test/controllers/SampleTestSpec.scala:22]: Is this file needed?

File: test/views/PensionSchemeViewSpec.scala [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    implicit val request = FakeRequest()
+    implicit val mess = messages(app)
+
+    "test 1" in {
+
+      val form = new PensionSchemeForm()
+      val viewModel = PensionSchemeViewModel(
+        DisplayMessage("value 1"),
+        DisplayMessage("header"),
+        //DisplayMessage("value 3")
```
Comment (kurtlogan) [inline test/views/PensionSchemeViewSpec.scala:48]: Don't comment out code, remove it instead

File: test/views/PensionSchemeViewSpec.scala [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+  running(_ => applicationBuilder()) { implicit app =>
+
+    val view = app.injector.instanceOf[PensionSchemeView]
+
+    implicit val request = FakeRequest()
+    implicit val mess = messages(app)
+
+    "test 1" in {
```
Comment (kurtlogan) [inline test/views/PensionSchemeViewSpec.scala:42]: Need a better test name

File: test/views/PensionSchemeViewSpec.scala [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        DisplayMessage("value 1"),
+        DisplayMessage("header"),
+        //DisplayMessage("value 3")
+      Call("GET", "/value")
+      )
+
+   val failingForm = form("value")
+
+      val result = view(failingForm, viewModel)
+      Jsoup.parse(result.toString()).body(). toString must include("header")
```
Comment (kurtlogan) [inline test/views/PensionSchemeViewSpec.scala:55]: no space between . and toString

File: test/views/PensionSchemeViewSpec.scala [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import play.api.test.FakeRequest
+import play.api.test.Helpers.running
+import viewmodels.models.PensionSchemeViewModel
+import viewmodels.DisplayMessage
+import views.html.PensionSchemeView
+import org.jsoup.Jsoup
+import play.api.mvc.Call
+
+
+class PensionSchemeViewSpec extends SpecBase {
```
Comment (kurtlogan) [inline test/views/PensionSchemeViewSpec.scala:32]: More tests needed in this file to cover success and failure cases

Comment (westwater) [reply test/views/PensionSchemeViewSpec.scala:32]: Added a success and failure case 👍

File: app/forms/DatePageFormProvider.scala (hunk 1 above)
Comment (westwater) [reply app/forms/DatePageFormProvider.scala:25]: Yeah it would be good to go through this

Comment (narendrapaduchuri) [reply app/forms/DatePageFormProvider.scala:25]: Yes, we can go through this

File: test/views/PensionSchemeViewSpec.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    implicit val request = FakeRequest()
+    implicit val mess = messages(application)
+
+    "PensionsSchemeView" - {
+
+      "view should render correctly" in {
+
+        val form = new PensionSchemeForm()
+        val viewModel = PensionSchemeViewModel(
```
Comment (kurtlogan) [inline test/views/PensionSchemeViewSpec.scala:47]: You need to test that every value passed into the view model is rendered on the page. So title should be tested, heading should be tested and the onSubmit should also be tested

File: test/views/PensionSchemeViewSpec.scala [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+          title = DisplayMessage("testTitle"),
+          heading = DisplayMessage("testHeading"),
+          Call("GET", "/value")
+        )
+
+        val failingForm = form("value").bind(Map("failing" -> "data"))
+
+        val result = view(failingForm, viewModel)
+        Jsoup.parse(result.toString()).body().getElementById("error-summary-title").text() mustBe "There is a problem"
```
Comment (kurtlogan) [inline test/views/PensionSchemeViewSpec.scala:69]: 2 error messages appear on the page, both should be tested

File: app/viewmodels/models/PensionSchemeViewModel.scala [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package viewmodels.models
+
+import play.api.mvc.Call
+import viewmodels.DisplayMessage
+
+case class PensionSchemeViewModel(
```
Comment (kurtlogan) [inline app/viewmodels/models/PensionSchemeViewModel.scala:22]: Might be a little clunky creating a view modeland having to pass `DisplayMessage`. Create a new apply method in the companion object with a signature `def apply(title: String, heading: String, onSubmit: Call): PensionSchemeViewModel` that applies the string values to `DisplayMessage`

File: app/forms/DatePageFormProvider.scala (hunk 1 above)
Comment (westwater) [reply app/forms/DatePageFormProvider.scala:25]: Had a chat and agreed to leave this for now

File: test/views/PensionSchemeViewSpec.scala (hunk 8 above)
Comment (westwater) [reply test/views/PensionSchemeViewSpec.scala:47]: Does the title get passed to the html? I thought it was for the browser tab

**Review by westwater — APPROVED** [review]
